File: backend/app/schemas/v3/statistics.py
"""统计报表 Schema"""

from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel


# ==================== 销售统计 ====================

class SalesStatItem(BaseModel):
    """销售统计项"""
    date: str = ""
    entity_id: Optional[int] = None
    entity_name: str = ""
    product_id: Optional[int] = None
    product_name: str = ""
    order_count: int = 0
    total_quantity: int = 0
    total_amount: float = 0
    total_cost: float = 0
    total_profit: float = 0
    profit_rate: float = 0  # 利润率


class SalesStatResponse(BaseModel):
    """销售统计响应"""
    items: List[SalesStatItem]
    summary: Dict[str, Any] = {}
    start_date: Optional[str] = None
    end_date: Optional[str] = None


# ==================== 采购统计 ====================

class PurchaseStatItem(BaseModel):
    """采购统计项"""
    date: str = ""
    entity_id: Optional[int] = None
    entity_name: str = ""
    product_id: Optional[int] = None
    product_name: str = ""
    order_count: int = 0
    total_quantity: int = 0
    total_amount: float = 0


class PurchaseStatResponse(BaseModel):
    """采购统计响应"""
    items: List[PurchaseStatItem]
    summary: Dict[str, Any] = {}
    start_date: Optional[str] = None
    end_date: Optional[str] = None


# ==================== 库存分析 ====================

class StockAnalysisItem(BaseModel):
    """库存分析项"""
    warehouse_id: int
    warehouse_name: str = ""
    product_id: int
    product_name: str = ""
    product_code: str = ""
    quantity: int = 0
    reserved_quantity: int = 0
    available_quantity: int = 0
    safety_stock: int = 0
    cost_price: float = 0
    stock_value: float = 0  # 库存金额 = 数量 × 成本价
    is_low_stock: bool = False  # 是否低库存


class StockAnalysisResponse(BaseModel):
    """库存分析响应"""
    items: List[StockAnalysisItem]
    total_stock_value: float = 0
    low_stock_count: int = 0
    out_of_stock_count: int = 0


# ==================== 库存预警 ====================

class StockWarningItem(BaseModel):
    """库存预警项"""
    warehouse_id: int
    warehouse_name: str = ""
    product_id: int
    product_name: str = ""
    product_code: str = ""
    quantity: int = 0
    safety_stock: int = 0
    shortage: int = 0  # 缺口 = 安全库存 - 当前库存
    warning_level: str = "low"  # low: 低库存, out: 缺货


class StockWarningResponse(BaseModel):
    """库存预警响应"""
    items: List[StockWarningItem]
    low_stock_count: int = 0
    out_of_stock_count: int = 0


# ==================== 仪表盘 ====================

class DashboardData(BaseModel):
    """仪表盘数据"""
    # 今日数据
    today_sales: float = 0
    today_sales_count: int = 0
    today_purchase: float = 0
    today_purchase_count: int = 0
    today_received: float = 0
    today_paid: float = 0
    
    # 本月数据
    month_sales: float = 0
    month_sales_count: int = 0
    month_purchase: float = 0
    month_purchase_count: int = 0
    month_profit: float = 0
    
    # 待处理
    pending_orders: int = 0
    draft_orders: int = 0
    
    # 账款
    total_receivable: float = 0
    total_payable: float = 0
    overdue_receivable: float = 0
    
    # 库存预警
    low_stock_count: int = 0
    out_of_stock_count: int = 0
    
    # 最近业务单
    recent_orders: List[Dict[str, Any]] = []
    
    # 销售趋势（最近7天/30天）
    sales_trend: List[Dict[str, Any]] = []


# ==================== 实体排行 ====================

class EntityRankItem(BaseModel):
    """实体排行项"""
    entity_id: int
    entity_name: str = ""
    entity_code: str = ""
    order_count: int = 0
    total_amount: float = 0
    total_profit: float = 0


class EntityRankResponse(BaseModel):
    """实体排行响应"""
    items: List[EntityRankItem]
    rank_type: str = "sales"  # sales: 销售排行, purchase: 采购排行


# ==================== 商品排行 ====================

class ProductRankItem(BaseModel):
    """商品排行项"""
    product_id: int
    product_name: str = ""
    product_code: str = ""
    total_quantity: int = 0
    total_amount: float = 0
    total_profit: float = 0


class ProductRankResponse(BaseModel):
    """商品排行响应"""
    items: List[ProductRankItem]
    rank_type: str = "sales"  # sales: 销量排行, profit: 利润排行

File: frontend/src/lib/api/v3/statistics.ts
/**
 * 统计报表 API
 */

import { API_BASE, getAuthHeaders, handleResponse } from './client';

export interface DashboardData {
  today_sales: number;
  today_sales_count: number;
  today_purchase: number;
  today_purchase_count: number;
  today_received: number;
  today_paid: number;
  month_sales: number;
  month_sales_count: number;
  month_purchase: number;
  month_purchase_count: number;
  month_profit: number;
  pending_orders: number;
  draft_orders: number;
  total_receivable: number;
  total_payable: number;
  overdue_receivable: number;
  low_stock_count: number;
  out_of_stock_count: number;
  recent_orders: Array<{
    id: number;
    order_no: string;
    order_type: string;
    type_display: string;
    status: string;
    status_display: string;
    final_amount: number;
    created_at: string;
  }>;
  sales_trend: Array<{ date: string; amount: number }>;
}

export interface SalesStatItem {
  date?: string;
  entity_id?: number;
  entity_name?: string;
  product_id?: number;
  product_name?: string;
  order_count: number;
  total_quantity: number;
  total_amount: number;
  total_cost?: number;
  total_profit?: number;
  profit_rate?: number;
}

export const statisticsApi = {
  getDashboard: async () => {
    const res = await fetch(`${API_BASE}/statistics/dashboard`, { headers: getAuthHeaders() });
    return handleResponse<DashboardData>(res);
  },
  
  getSales: async (params?: {
    group_by?: string;
    start_date?: string;
    end_date?: string;
    limit?: number;
  }) => {
    const query = new URLSearchParams();
    if (params?.group_by) query.set('group_by', params.group_by);
    if (params?.start_date) query.set('start_date', params.start_date);
    if (params?.end_date) query.set('end_date', params.end_date);
    if (params?.limit) query.set('limit', params.limit.toString());
    
    const res = await fetch(`${API_BASE}/statistics/sales?${query}`, { headers: getAuthHeaders() });
    return handleResponse<{ items: SalesStatItem[]; summary: Record<string, any>; start_date?: string; end_date?: string }>(res);
  },
  
  getPurchase: async (params?: {
    group_by?: string;
    start_date?: string;
    end_date?: string;
    limit?: number;
  }) => {
    const query = new URLSearchParams();
    if (params?.group_by) query.set('group_by', params.group_by);
    if (params?.start_date) query.set('start_date', params.start_date);
    if (params?.end_date) query.set('end_date', params.end_date);
    if (params?.limit) query.set('limit', params.limit.toString());
    
    const res = await fetch(`${API_BASE}/statistics/purchase?${query}`, { headers: getAuthHeaders() });
    return handleResponse<{ items: SalesStatItem[]; summary: Record<string, any>; start_date?: string; end_date?: string }>(res);
  },
  
  getEntityRank: async (params?: { rank_type?: string; start_date?: string; end_date?: string; limit?: number }) => {
    const query = new URLSearchParams();
    if (params?.rank_type) query.set('rank_type', params.rank_type);
    if (params?.start_date) query.set('start_date', params.start_date);
    if (params?.end_date) query.set('end_date', params.end_date);
    if (params?.limit) query.set('limit', params.limit.toString());
    
    const res = await fetch(`${API_BASE}/statistics/entity-rank?${query}`, { headers: getAuthHeaders() });
    return handleResponse<{ items: Array<{ entity_id: number; entity_name: string; entity_code: string; order_count: number; total_amount: number }>; rank_type: string }>(res);
  },
  
  getProductRank: async (params?: { rank_type?: string; start_date?: string; end_date?: string; limit?: number }) => {
    const query = new URLSearchParams();
    if (params?.rank_type) query.set('rank_type', params.rank_type);
    if (params?.start_date) query.set('start_date', params.start_date);
    if (params?.end_date) query.set('end_date', params.end_date);
    if (params?.limit) query.set('limit', params.limit.toString());
    
    const res = await fetch(`${API_BASE}/statistics/product-rank?${query}`, { headers: getAuthHeaders() });
    return handleResponse<{ items: Array<{ product_id: number; product_name: string; product_code: string; total_quantity: number; total_amount: number; total_profit: number }>; rank_type: string }>(res);
  }
};

